text/0000-confidential-token-ids.md

```diff
@@ -0,0 +1,270 @@
+
+- To recover the `token_id` from the `masked_token_id`, we reverse the process (if the bytes field is length = 4, otherwise `token_id = 0`).
+
+    ```
+    unmasked_token_id = masked_token_id ^ Blake2b(token_id_tag | shared_secret)
+    ```
+
+### Proof of Opening: Proof of Knowledge of Uniform Token ID in the Transaction
+
+We would like to be able to prove that all transaction inputs and outputs are using the same `token_ids`. We do this using a non-interactive, zero knowledge Proof of Opening that the sender has the capability to unwrap the sum of output commitments in the transaction, in conjunction with the homomorphic encryption property of [Pedersen Commitments](https://www.cs.cornell.edu/courses/cs754/2001fa/129.PDF), namely that addition on the commitments preserves the additive relationship of the pre-committed values, as long as they are using the same group generator.
+
+- To prove each TXO's commitment to the `amount` value, we use a Pedersen commitment, consisting of a group generator (`H_i`) raised to the value of the `amount`, with a blinding factor to prevent brute-forcing the full range of values (`2^64`). We compute the blinding factor as the blinding group generator raised to a hash of the shared secret for the TXO.
+
+    ```
+    blinding_factor = Blake2B(blinding_tag | shared_secret)
+    commitment = H_i * Scalar::from(amount) + G * blinding_factor
+    ```
+
+    - Note, the group generators for Token Types are computed using hashing to a curve so that they are "orthogonal" for computable adversaries, meaning there is no discernable, nontrivial linear relationship between base points. See our [transaction core group generators reference](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/transaction/core/src/ring_signature/mod.rs#L29). 
+    - Note also, the blinding factor base point does not need to be unique per TokenType, so we use the same blinding factor base point for all TokenTypes.
+    - Note finally, the Proof of Opening requires all commitments in a transaction were generated using the same group generator, base point `H_i`. If the commitments in input and/or output TXOs in a transaction are constructed with different group generators, and the sender is able to unwrap the sum as specified in the protocol below, it implies that the sender knows a nontrivial linear relationship among the `H_i`, which contradicts the assumption that the `H_i` are orthogonal. 
+
+The non-interactive zero knowledge Proof of Opening protocol is the following: (See [Zero Knowledge Proofs and Commitment Schemes, Page 27](https://www.cs.purdue.edu/homes/ninghui/courses/555_Spring12/handouts/555_Spring12_topic23.pdf))
+
+
+1. The prover establishes a value, `d`, calculated by hashing their secret values from the commitment (the `token_id` and the `blinding_factor`). Implementation note, we may use [Merlin](https://docs.rs/merlin/latest/merlin/) rather than Blake2B.
+
+    ```
+    y = Blake2B(proof_tag_t | token_id)
```

Comment:
If `token_id` is only 32 bits, this is not secure (`u` can be trivially guess-and-checked).

Reply:
Yeah this is not what I did in the draft PR, my draft code looks like this:

```

        // Compute prover secrets y and s, and the commitment based on these.
        //
        // These are unpredictable assuming that the blinding factor is unpredictable
        // and that Merlin adequately implements Fiat Shamir.
        let (y, s) = {
            let mut transcript =
                MerlinTranscript::new(PROOF_OF_OPENING_SECRET_DOMAIN_TAG.as_bytes());
            value.append_to_transcript(b"x", &mut transcript);
            blinding.append_to_transcript(b"r", &mut transcript);

            let mut y_bytes = [0u8; 32];
            let mut s_bytes = [0u8; 32];

            transcript.challenge_bytes(b"y", &mut y_bytes);
            transcript.challenge_bytes(b"s", &mut s_bytes);

            (Scalar::from_bits(y_bytes), Scalar::from_bits(s_bytes))
        };
```

so that the blinding factor feeds into both prover secrets and can't be brute forced

Reply:
We decided to remove the proof of opening, so marking this conversation resolved!